src/posix-socket.c

```diff
@@ -0,0 +1,612 @@
+    }
+    size_t sz;
+    struct sockaddr* sockaddr = (struct sockaddr*)JS_GetUint8Array(ctx, &sz, argv[0]);
+    CHECK_ARG_RET(ctx, sockaddr, 0, "Uint8Array");
+    int ret = bind(s->sock, sockaddr, sz);
+    RET_THROW_ERRNO(ctx, ret == 0);
+
+    return JS_UNDEFINED;
+}
+
+static JSValue tjs_sock_accept(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+    tjs_sock_t* s = JS_GetOpaque(this_val, tjs_sock_classid);
+    CHECK_ARG_RET(ctx, s, -1, tjs_sock_class.class_name);
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+
+    socklen_t sz = sizeof(struct sockaddr);
+    struct sockaddr* sockaddr = js_malloc(ctx, sz);
+    int ret = accept(s->sock, sockaddr, &sz);
+    if (ret < 0) {
+        js_free(ctx, sockaddr);
+        return THROW_STRERROR();
+    }
+    
+    JSValue newSock = tjs_sock_new_from_fd(ctx, ret);
+    JS_SetPropertyStr(ctx, newSock, "_sockaddr", TJS_NewUint8Array(ctx, (uint8_t*)sockaddr, sz));
+    return newSock;
+}
+
+static JSValue tjs_sock_connect(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+    tjs_sock_t* s = JS_GetOpaque(this_val, tjs_sock_classid);
+    CHECK_ARG_RET(ctx, s, -1, tjs_sock_class.class_name);
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+
+    size_t sz;
+    struct sockaddr* sockaddr = (struct sockaddr*)JS_GetUint8Array(ctx, &sz, argv[0]);
+    CHECK_ARG_RET(ctx, sockaddr, 0, "Uint8Array");
+    
+    int ret = connect(s->sock, sockaddr, sz);
+    RET_THROW_ERRNO(ctx, ret == 0);
+    
+    return JS_UNDEFINED;
+}
+
+static JSValue tjs_sock_setsockopt(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+    tjs_sock_t* s = JS_GetOpaque(this_val, tjs_sock_classid);
+    CHECK_ARG_RET(ctx, s, -1, tjs_sock_class.class_name);
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+
+    int level, optname;
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, &level, argv[0]), 0, "positive integer");
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, &optname, argv[1]), 1, "positive integer");
+    size_t optlen;
+    void* optval = JS_GetUint8Array(ctx, &optlen, argv[2]);
+    CHECK_ARG_RET(ctx, optval, 2, "Uint8Array");
+    
+    int ret = setsockopt(s->sock, level, optname, optval, optlen);
+    RET_THROW_ERRNO(ctx, ret == 0);
+    
+    return JS_UNDEFINED;
+}
+
+static JSValue tjs_sock_getsockopt(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+    tjs_sock_t* s = JS_GetOpaque(this_val, tjs_sock_classid);
+    CHECK_ARG_RET(ctx, s, -1, tjs_sock_class.class_name);
+    if(s->closed){
+        return JS_ThrowInternalError(ctx, "Socket closed");
+    }
+
+    int level, optname;
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, &level, argv[0]), 0, "positive integer");
+    CHECK_ARG_RET(ctx, !JS_ToUint32(ctx, &optname, argv[1]), 1, "positive integer");
+    
+    socklen_t optlen = 50; // TODO: adjust value 
```

Comment:
what options are so large?

Reply:
The largest one I could find for SOL_SOCKET is `struct sockaddr_storage` for `SO_PEERNAME` which is 128byte. But there could be larger ones for more specific Socket types. E.g. Bluetooth socket or maybe some more specific TCP/IP options getable via getsockopt. So now I just let the user pass an optional size, otherwise I will use 128 as the default.

Reply:
FYI: Just have a case where a socket option can get arbitrarily long, `SOL_SOCKET, SO_ATTACH_FILTER` (BPF Packet filter).